discovery-provider/src/eth_indexing/event_scanner.py

```diff
@@ -0,0 +1,434 @@
+
+    def process_event(self, block_when: datetime.datetime, event: TransferEvent) -> str:
+        """Record a ERC-20 transfer in our database."""
+        # Events are keyed by their transaction hash and log index
+        # One transaction may contain multiple events
+        # and each one of those gets their own log index
+
+        # event_name = event.event # "Transfer"
+        log_index = event["logIndex"]  # Log index within the block
+        # transaction_index = event.transactionIndex  # Transaction index within the block
+        txhash = event["transactionHash"].hex()  # Transaction hash
+        block_number = event["blockNumber"]
+
+        # Convert ERC-20 Transfer event to our internal format
+        args = event["args"]
+        transfer = {
+            "from": args["from"],
+            "to": args["to"],
+            "value": args["value"]
+            / 1e18,  # divide by 10^18 to get correct transfer value
+            "timestamp": block_when,
+        }
+
+        # add user ids from the transfer event into the balance refresh queue
+        transfer_event_wallets = [transfer["from"].lower(), transfer["to"].lower()]
+        with self.db.scoped_session() as session:
+            result = (
+                session.query(User.user_id)
+                .outerjoin(AssociatedWallet, User.user_id == AssociatedWallet.user_id)
+                .filter(User.is_current == True)
+                .filter(AssociatedWallet.is_current == True)
+                .filter(AssociatedWallet.is_delete == False)
+                .filter(
+                    or_(
+                        User.wallet.in_(transfer_event_wallets),
+                        AssociatedWallet.wallet.in_(transfer_event_wallets),
+                    )
+                )
+                .all()
+            )
+            user_ids = [user_id for [user_id] in result]
+            enqueue_balance_refresh(self.redis, user_ids)
+
+        # Return a pointer that allows us to look up this event later if needed
+        return f"{block_number}-{txhash}-{log_index}"
+
+    def scan_chunk(self, start_block, end_block) -> Tuple[int, list]:
+        """Read and process events between to block numbers.
+
+        Dynamically decrease the size of the chunk if the case JSON-RPC server pukes out.
+
+        :return: tuple(actual end block number, when this block was mined, processed events)
+        """
+
+        block_timestamps = {}
+        get_block_timestamp = self.get_block_timestamp
+
+        # Cache block timestamps to reduce some RPC overhead
+        # Real solution might include smarter models around block
+        def get_block_when(block_num):
+            if block_num not in block_timestamps:
+                block_timestamps[block_num] = get_block_timestamp(block_num)
+            return block_timestamps[block_num]
+
+        all_processed = []
+
+        # Callable that takes care of the underlying web3 call
+        def _fetch_events(_start_block, _end_block):
```

Comment:
nit: the underscore prefixed variable names are a bit weird, we don't have this anywhere else in the codebase